velox/functions/sparksql/Comparisons.cpp

```diff
@@ -0,0 +1,319 @@
+                           ->template mutableRawValues<uint64_t>();
+      auto constant = args[1]->asUnchecked<ConstantVector<bool>>()->valueAt(0);
+      rows.applyToSelected([&](vector_size_t i) {
+        flatResult->set(i, cmp(bits::isBitSet(rawValues, i), constant));
+      });
+    } else {
+      // Fast path if one or more arguments are encoded.
+      exec::DecodedArgs decodedArgs(rows, args, context);
+      auto decoded0 = decodedArgs.at(0);
+      auto decoded1 = decodedArgs.at(1);
+      rows.applyToSelected([&](vector_size_t i) {
+        flatResult->set(
+            i, cmp(decoded0->valueAt<bool>(i), decoded1->valueAt<bool>(i)));
+      });
+    }
+  }
+};
+
+template <template <typename> class Cmp>
+std::shared_ptr<exec::VectorFunction> makeImpl(
+    const std::string& functionName,
+    const std::vector<exec::VectorFunctionArg>& args) {
+  VELOX_CHECK_EQ(args.size(), 2);
+  for (size_t i = 1; i < args.size(); i++) {
+    VELOX_CHECK(*args[i].type == *args[0].type);
+  }
+  switch (args[0].type->kind()) {
+#define SCALAR_CASE(kind)                            \
+  case TypeKind::kind:                               \
+    return std::make_shared<ComparisonFunction<      \
+        Cmp<TypeTraits<TypeKind::kind>::NativeType>, \
+        TypeKind::kind>>();
+    SCALAR_CASE(TINYINT)
+    SCALAR_CASE(SMALLINT)
+    SCALAR_CASE(INTEGER)
+    SCALAR_CASE(BIGINT)
+    SCALAR_CASE(HUGEINT)
+    SCALAR_CASE(REAL)
+    SCALAR_CASE(DOUBLE)
+    SCALAR_CASE(VARCHAR)
+    SCALAR_CASE(VARBINARY)
+    SCALAR_CASE(TIMESTAMP)
+#undef SCALAR_CASE
+    case TypeKind::BOOLEAN:
+      return std::make_shared<BoolComparisonFunction<
+          Cmp<TypeTraits<TypeKind::BOOLEAN>::NativeType>>>();
+    default:
+      VELOX_NYI(
+          "{} does not support arguments of type {}",
+          functionName,
+          args[0].type->kind());
+  }
+}
+
+template <TypeKind kind>
+void applyTyped(
+    const SelectivityVector& rows,
+    std::vector<VectorPtr>& args,
+    DecodedVector* decoded0,
+    DecodedVector* decoded1,
+    exec::EvalCtx& context,
+    FlatVector<bool>* flatResult) {
+  using T = typename TypeTraits<kind>::NativeType;
+
+  Equal<T> equal;
+  if (!args[0]->mayHaveNulls() && !args[1]->mayHaveNulls()) {
+    // When there is no nulls, it reduces to normal equality function
+    rows.applyToSelected([&](vector_size_t i) {
+      flatResult->set(
+          i, equal(decoded0->valueAt<T>(i), decoded1->valueAt<T>(i)));
+    });
+  } else {
+    // (isnull(a) AND isnull(b)) || (a == b)
+    // When DecodedVector::nulls() is null it means there are no nulls.
+    auto* rawNulls0 = decoded0->nulls();
+    auto* rawNulls1 = decoded1->nulls();
+    rows.applyToSelected([&](vector_size_t i) {
+      auto isNull0 = rawNulls0 && bits::isBitNull(rawNulls0, i);
+      auto isNull1 = rawNulls1 && bits::isBitNull(rawNulls1, i);
+      flatResult->set(
+          i,
+          (isNull0 || isNull1)
```

Comment:
I want to make sure this is spark semantics , if any of values is Null, then its only true if both are null , right ?

Reply:
Yes, in Spark, it will return true if both sides are null and return false if only one side is null.